### DISCUSSION TITLE
Is it intentional for extension methods to not be allowed as top-level functions?

### DISCUSSION THREAD

**Discussion by IanKemp** [dotnet/csharplang#4526] [Q&A] [answered]
Prompted by this Stack Overflow question: https://stackoverflow.com/q/66581096/70345 (please read the answers for additional context/speculation).

The tl;dr is that Roslyn does not allow defining extension methods as top-level functions. [The relevant spec](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/top-level-statements.md) does not appear to define any special behaviour around top-level functions that would preclude them from being extension methods, however the [extension methods spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/classes#extension-methods) states that:

> Extension methods can only be declared in non-generic, non-nested static classes.

Since a top-level function is not technically defined in a class, [**even though it's compiled to a method inside a static `Program` class**](https://stackoverflow.com/a/66581366/70345), is the required declaration of extension methods in a static class the overriding factor here? In other words, does the extension methods spec have "higher precedence" than the top-level statements one, and/or is there language missing from one or both of these specs regarding their potential for interplay?

Alternatively, should Roslyn be allowing extension methods as top-level functions, i.e. is there a bug in Roslyn's implementation of the top-level statements spec?

**Comment by Joe4evr** [comment]
[Yes.](https://github.com/dotnet/csharplang/blob/a3d67d141501101af9dbb9e8192ca8b4c717e316/spec/namespaces.md#using-static-directives) (emphasis mine):
> Within member declarations in a compilation unit or namespace body that contains a *using_static_directive*, the accessible nested types and static members **(except extension methods)** contained directly in the declaration of the given type can be referenced directly.
> [...]
> A *using_static_directive* specifically does not import extension methods directly as static methods, but makes them available for extension method invocation.

**Reply by canton7** [reply]
Is `using static` relevant here?

**Comment by HaloFour** [comment, marked as answer]
IIRC top-level functions are really just local functions declared within the implicit `Main` method, and local functions can't be extension methods.

**Reply by IanKemp** [reply]
> top-level functions are really just local functions declared within the implicit Main method

That's not how Roslyn compiles them, though, according to SharpLab: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQEsB2MIIA4D2ANrBACarIwJYDmANGSANQA+AAgIwB0XAnAAoARAAkIRIgQAEAdQIIiZIQEoA3AFgAUFvYAWKQBECAZQIBbCABUAFnQHKtAby1TXUrr06DR4ybPmKKhqaAL5AA===

**Reply by HaloFour** [reply]
That's exactly how Roslyn compiles all local functions.  However, that is an implementation detail and what gets emitted depends on what you do with the local function.  It will be emitted differently if you enclose local context, or take a delegate to the local function.

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQEsB2MIIA4D2ANrBACarIwJYDmANGSANQA+AAgEwDMABAN4BYAFAje43l04CxEuewCMAOkUBOABQAiABIQiRArwDqBBETKaAlAG4RsuePYAWXgBECAZQIBbCABUACzp1S3sHIWEHKMllNS1dfUMTMwsbMLkAXzCs4QygA===

The only differences being that the name of the emitted method has changed because it's based on the auto-generated `Main` method and the `CompilerGenerated` attribute is applied to the entire class rather than just the emitted method.

**Reply by HaloFour** [reply]
That said I don't think that's a reason that top-level functions _can't_ be extension methods either on their own or by allowing local functions to also be considered extension methods.  I've had cases where the latter would have been useful.

**Reply by connellsharp** [reply]
I'm beginning to think this is the correct way to think about it. In two steps:
1. Top-level functions are converted to local functions within the `Main` method.
2. Those local functions are then lifted into the `Program` class, as part of the normal implementation of local functions.

So these "top-level methods" should be thought of as local functions when developing, and the same restrictions apply. E.g. you can't declare local extension methods.

If top-level extension methods were to be included in the language in future, it feels like it would make sense to do it as part of a more general feature that allows local extension methods. Which would mean top-level extensions are inaccessible from outside of `Program.Main`.

**Reply by 333fred** [reply]
This is exactly the correct way to think about them. Top-level statements (including top-level function declarations) are wrapped in an implicit Main method. What happens after that is the standard way methods are compiled, including local functions inside a method.

**Reply by IanKemp** [reply]
I'm still confused. Given the following class declaration (which is essentially the same as the top-level statements we've been discussing):

```csharp
static class Program
{
    public static void Main()
    {
        static int ToInt(this string numberString) => int.Parse(numberString);
        
        "42".ToInt();
    }
}
```

I would (based on my understanding of how local functions are lifted into class methods) expect Roslyn to resolve it to the following:

```csharp
static class Program
{
    static int ToInt(this string numberString) => int.Parse(numberString);
    
    public static void Main()
    {
        "42".ToInt();
    }
}
```

And indeed, if I paste the latter program into SharpLab, **it compiles as expected**:

```csharp
[Extension]
internal static class Program
{
    [Extension]
    private static int ToInt(string numberString)
    {
        return int.Parse(numberString);
    }

    public static void Main()
    {
        ToInt("42");
    }
}
```

So why doesn't the first one?

**Reply by HaloFour** [reply]
How they are lowered is an implementation detail and as with normal local functions it will change depending on what else you do in the method.  The compiler doesn't consider those implementation details when it considers whether or not a local function is eligible to be considered as an extension method.  Same as how local functions don't participate with overloads of instance members in the same class.

**Reply by canton7** [reply]
>  The compiler doesn't consider those implementation details when it considers whether or not a local function is eligible to be considered as an extension method

(I made that abundantly clear in [my StackOverflow answer](https://stackoverflow.com/a/66581581/1086121) to this question).

Extension methods are specified as only being valid when **declared** in static (non-generic, non-nested) classes. The spec says this. It doesn't matter what the compiler does afterwards, what matters is how it's **declared**. The fact that the compiler ends up lifting top-level methods into a generated class is completely irrelevant.

Yes, the spec could be changed, but right now the compiler will only accept an extension method if it's **declared** in a static class.
